Feature: add title to action execution errors to improve user experience

sumitsum opened the pull request:
## Description
- add title to action execution errors to improve user experience
  - all errors in AppsmithPluginErrors
  - action execution related errors in AppsmithError : `NO_CONFIGURATION_FOUND_IN_DATASOURCE`, `INVALID_ACTION`, `INVALID_DATASOURCE`, `INVALID_DATASOURCE_CONFIGURATION` - as these are the likely errors in an action execution flow (excluding plugin specific flow) that might result from faulty action / datasource configuration by a user.
  - title is returned as part of ActionExecutionResult. 
  - title is set in ActionExecutionResult on failure during action execution.
  - In response to review comment - added a new BaseException class, so that both AppsmithException and AppsmithPluginException extend this base class. Also, refactored code to introduce setErrorInfo() function in ActionExecutionResult class to set its attributes. 
- Some unrelated cleanup: 
    - catch PoolInitializationException for Postgres plugin and return AppsmithPuginException
    - catch MongoTimeoutException and return AppsmithPluginException

Fixes 3660 
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- JUnit TCs
- Manual
![Screenshot 2021-04-05 at 4 06 31 PM](https://user-images.githubusercontent.com/1757421/113662665-2dbd3e00-96c6-11eb-90f0-1e60d4056d4d.png)
![Screenshot 2021-04-05 at 6 22 35 PM](https://user-images.githubusercontent.com/1757421/113662675-3150c500-96c6-11eb-9393-1ac7102c3af8.png)
![Screenshot 2021-04-06 at 9 19 54 AM](https://user-images.githubusercontent.com/1757421/113662677-3281f200-96c6-11eb-81c9-7c1e8d1500d5.png)


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

nidhi-nair commented:
@sumitsum  Is it possible to abstract away details of how an error state is stored in ActionExecutionResult? Setting error codes and titles separately increases the responsibility of the developer for every new error that needs to be logged. Instead, if we use a method like `ActionExecutionResult.setErrorState(AppsmithError)`, then all of the error handling translation can happen within it. This will make it easier to modify this structure in the future as well.

sumitsum commented:
> @sumitsum Is it possible to abstract away details of how an error state is stored in ActionExecutionResult? Setting error codes and titles separately increases the responsibility of the developer for every new error that needs to be logged. Instead, if we use a method like `ActionExecutionResult.setErrorState(AppsmithError)`, then all of the error handling translation can happen within it. This will make it easier to modify this structure in the future as well.

@nidhi-nair I have added a method named `setErrorInfo` to`ActionExecutionResult` to handle state assignment.

sumitsum commented:
> @sumitsum Is it possible to abstract away details of how an error state is stored in ActionExecutionResult? Setting error codes and titles separately increases the responsibility of the developer for every new error that needs to be logged. Instead, if we use a method like `ActionExecutionResult.setErrorState(AppsmithError)`, then all of the error handling translation can happen within it. This will make it easier to modify this structure in the future as well.

@nidhi-nair  I have refactored code to introduce `setErrorInfo()` method in `ActionExecutionResult `class to set its attributes. Also, added a new `BaseException` class, so that both` AppsmithException` and `AppsmithPluginException `extend this base class. This was important since `ActionExecutionResult` class is defined in `com.appsmith.external `package and `AppsmithException` class is in `com.appsmith.server` package and a dependency from `server` package to `external` package could not be created.
